src/build.d

```diff
@@ -1301,23 +1274,23 @@ Params:
 Returns:
     `true` if the target is up to date
 */
-auto isUpToDate(string[] targets, string[][] sources...)
+auto isUpToDate(R, S)(R targets, S sources)
 {
     if (force)
         return false;
-
+    auto oldestTargetTime = SysTime.max;
     foreach (target; targets)
     {
-        auto sourceTime = target.timeLastModified.ifThrown(SysTime.init);
-        // if a target has no sources, it only needs to be built once
-        if (sources.empty || sources.length == 1 && sources.front.empty)
-            return sourceTime > SysTime.init;
-        foreach (arg; sources)
-            foreach (a; arg)
-                if (sourceTime < a.timeLastModified.ifThrown(SysTime.init + 1.seconds))
-                    return false;
+        const time = target.timeLastModified.ifThrown(SysTime.init);
+        if (time == SysTime.init)
+            return false;
+        oldestTargetTime = min(time, oldestTargetTime);
+    }
+    foreach (source; sources)
+    {
+        if (source.timeLastModified.ifThrown(SysTime.init) > oldestTargetTime)
+            return false;
     }
-
     return true;
```

Comment:
Could collapse the loop using `all`:

```
return sources.all!(source => source.timeLastModified.ifThrown(SysTime.init) <= oldestTargetTime);
```

Reply:
Doesn't allow short-circuiting. Some dependencies have alot of dependencies and we want to avoid reading file timestamps if we don't have to.

Reply:
Oh wait, `all` does allow short circuiting. Very nice, I'll use this.